maskrcnn_benchmark/structures/segmentation_mask.py

```diff
@@ -8,68 +11,213 @@
+        '''
+            Arguments:
+                masks: Either torch.tensor of [num_instances, H, W]
+                    or list of torch.tensors of [H, W] or
+                    BinaryMaskList.
+                size: absolute image size, width first
+        '''
+
+        if isinstance(masks, torch.Tensor):
+            pass
+        elif isinstance(masks, (list, tuple)):
+            masks = torch.stack(masks, dim=2)
+        elif isinstance(masks, BinaryMaskList):
+            masks = masks.masks
+
+        if len(masks.shape) == 2:
+            masks = masks[None]
+        assert len(masks.shape) == 3
+
+
+        assert masks.shape[1] == size[1],\
+            '%s != %s'%(masks.shape[1], size[1])
+        assert masks.shape[2] == size[0],\
+            '%s != %s'%(masks.shape[2], size[0])
+
+
+        self.masks = masks.clone()
+        self.size = tuple(size)
 
-    def __init__(self, masks, size, mode):
-        self.masks = masks
-        self.size = size
-        self.mode = mode
 
     def transpose(self, method):
-        if method not in (FLIP_LEFT_RIGHT, FLIP_TOP_BOTTOM):
-            raise NotImplementedError(
-                "Only FLIP_LEFT_RIGHT and FLIP_TOP_BOTTOM implemented"
+        dim = 1 if method == FLIP_TOP_BOTTOM else 2
+        # print(dim)
+        # print(self.masks)
+        # print(self.masks.flip(dim))
+        flipped_masks = self.masks.flip(dim)
+        return BinaryMaskList(flipped_masks, self.size)
+
+
+    def crop(self, box):
+        assert isinstance(box, (list, tuple)), str(type(box))
+
+        # box is assumed to by xyxy
+        current_width, current_height = self.size
+        xmin, ymin, xmax, ymax = map(round, box)
+        assert xmin >= 0 and xmax < current_width
+        assert ymin >= 0 and ymax < current_height
+        assert xmin < xmax and ymin < ymax
+
+        width, height = xmax - xmin, ymax - ymin
+        #width = max(width, 1)
+        #height = max(height, 1)
+
+        if (xmin >= xmax or ymin >= ymax):
+            print(box)
+
+        cropped_masks = self.masks[:, ymin:ymax, xmin:xmax]
+        cropped_size = width, height
+        return BinaryMaskList(cropped_masks, cropped_size)
+
+
+    def resize(self, size):
+        try:
+            iter(size)
+        except TypeError:
+            assert isinstance(size, (int, float))
+            size = size, size
+        width, height = map(int, size)
+
+        assert width > 0
+        assert height > 0
+
+        # Height comes first here!
+        resized_masks = torch.nn.functional.interpolate(
+            input=self.masks[None].float(),
+            size=(height, width),
+            mode='bilinear',
+            align_corners=False
+        )[0].type_as(self.masks)
+        resized_size = width, height
+        return BinaryMaskList(resized_masks, resized_size)
+
+
+
+    def convert_to_polygon(self):
+        contours = self._findContours()
+        return PolygonList(contours, self.size)
```

Comment:
This is not generally true. Consider a  donut shape mask, the contours would fail to represent the real shape.

Reply:
Please refer to [L139](https://github.com/facebookresearch/maskrcnn-benchmark/pull/473/files/d79829d0438636447e0137d9df7bcb19cf9d307d#diff-928af5178eceaef7d662fe22c85f439aR139):
```
contour, hierarchy = cv2.findContours(
  mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_TC89_L1
)
```
This solves the donut problem:
- [binary mask](https://pasteboard.co/I2c1EbA2.png)
- polygon [using cv2.RETR_TREE](https://pasteboard.co/I2c2bCr.png)
- polygon [using cv2.RETR_EXTERNAL](https://pasteboard.co/I2c2Lby.png)

Reply:
Have you tested with a binary mask with multiple holes? I am not sure whether it is generalizable to all cases.

Reply:
can you give a counterexample, where you cannot select the outermost contour?

Reply:
@botcs Can you convert the contours of a donut-shape object or object with holes found by cv2 back to the original binary mask via cocomask API? At least in my experience this is not trivial.

Besides, the function you provided has a minor problem with my opencv version. 
It should be `_, contour, hierarchy = cv2.findContours(
  mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_TC89_L1
)`

Reply:
I will check it out, could you please give me a polygon template to use?

About the opencv issue: previously I have came across an aux function that supports both versions, 
Personally, I would suggest updating it to the latest version which does not change the passed mask argument, therefore it is not returning it either since the whole project is using nightly-builds and there is little to no support on backwards compatibility.

Reply:
Just use the polygon testcase in my original unitest，where the second polygon is actually the hole of the object.

Reply:
I removed that part from the test since it did not change anything of the binary mask, which makes me think that the conversion is performed on both polygon instances but there are no signs why would be the inner hole's mask subtracted from the outer.

One way to come around this is to convert each polygon entities of the PolygonInstance using a N-ary XOR that would translate to 1 IFF the pixel is covered by a single polygon.

What do you think?

Reply:
I think this is OK, but maybe a warning or comments should be noted to users that if their dataset contains objects with holes, this conversion is somehow problematic.

Reply:
I have now looked into the details of this `polygon -> mask` function carried out by COCO API, and the following happens: each polygon entity (parts of a polygon instance) is converted to RLE and the RLE is converted to mask.

This whole unnecessary overhead could be simplified by using [`cv2.fillPoly`](https://docs.opencv.org/3.0-beta/modules/imgproc/doc/drawing_functions.html?highlight=fillpoly#void%20fillPoly(Mat&%20img,%20const%20Point**%20pts,%20const%20int*%20npts,%20int%20ncontours,%20const%20Scalar&%20color,%20int%20lineType,%20int%20shift,%20Point%20offset)) which could handle the output of the `Mask->Polygon` conversion in a very neat way. This could allow a probably *lossless* `Mask<->Polygon` bi-directional conversion.

What is your opinion?

Reply:
@botcs I'd prefer not to have a dependency on OpenCV in the core training loop (I've seen reports in the past that it didn't mixed well with pytorch multiprocessing, even though it might not apply here).

About lossless conversion from Mask to Polygon, I don't think this is actually possible, because there are discretization artifacts introduced when we go from a Polygon to a Mask that can't be exactly recovered. That being said, it would definitely be nice to have a way to convert back from mask to polygon, but in this case, given that polygons take less space, wouldn't it be recommended to (almost) always use polygons instead?

Reply:
@fmassa I agree with the dependency issue. Should I import it only when the conversion is called and raise a warning, or remove the feature?

I agree, but apart from corner cases, in worst case the boundary pixels' coordinates would serve as the polygon representation. In reverse, if no shrinking or warping has been applied I think conversion is completely reversible.

> given that polygons take less space, wouldn't it be recommended to (almost) always use polygons instead?

in theory yes, in practice I think it depends how often does the conversion happen. In synthetic data there are ridiculously rough surfaces which would be expensive to store and as long as polygon operations are not parallelized, masks **could be** dealt with more easily. 

To sum up: In the current version polygon conversion returns the outer boundary only, no polygons with holes can be represented; on the other side we don't have parallelized binary mask transformations.